### PR TITLE
Packaging fixes [v2]

### PR DESCRIPTION
**Pull request by clebergnu**
This PR replaces the custom usage of "data_files" parameter to setuptools, containing locations that start with `/usr/share/`, for Python standard way of packaging data files.

This means there are a lot of files moved around in this PR, but untouched with regards to their contents.

It fixes a long standing bug that was not being covered by our deployment checks on Cirrus CI, and that we've finally found out that would only affect installations via wheels (other pip and setup.py installations worked fine).

---

Changes from v1 (#2806):
 * Removed unused imports

### CODE REVIEW

**Comment by pevogam** [comment]
Hi @clebergnu do you mind providing a reference for the suggested standard way of packaging files?

**Comment by clebergnu** [comment]
> Hi @clebergnu do you mind providing a reference for the suggested standard way of packaging files?

Oh, sure! Should have done that in the commit message.  Look for "data_files" here:

https://setuptools.readthedocs.io/en/latest/references/keywords.html

And the referenced issue (https://github.com/avocado-framework/avocado-vt/issues/2193) shows the problem in action.

**Review by willianrampazzo — APPROVED** [review]

**Comment by clebergnu** [comment]
@ldoktor I see you're "around", so if you happen to have spare cycles and can review this, it'd be appreciated!

**Comment by clebergnu** [comment]
@pevogam have you had a chance to review/test this? Thanks!

**Comment by pevogam** [comment]
> @pevogam have you had a chance to review/test this? Thanks!

Sure, will do so today. As this PR needs some extra testing, I need some more time to set up a proper environment and couldn't review right away.

**Review by pevogam — COMMENTED** [review]
Hi @clebergnu, basic tests are ok and more are running now so I am ready to comment. All the changes here look completely reasonable but I am wondering something about the change in the data files path.  In particular, the current standard python way seems to be:

1) too monolithic placing various data files in the same location where the library code is placed
2) disregarding of system wide paths and common practices like using `/usr/share` for additional files and `/misc` for configuration for instance
3) created a clearer and clearer system-within-the-system feel of python packages being differently named and starting to use their own paths for nearly everything

At my first expectation when checking the modules under python's site-packages directory won't be about all possible files of the packages other than the actual importable modules. What is your opinion about this? I know some of my concerns are wider than just VT so specifically about the local RPM, could we at least keep the RPM more system-friendly by using the system-wide paths? You already mention it could customize paths on top so I see no reason why this shouldn't be possible.

**Review by beraldoleal — APPROVED** [review]
LGTM

**Comment by pevogam** [comment]
> At my first expectation when checking the modules under python's site-packages directory won't be about all possible files of the packages other than the actual importable modules. What is your opinion about this? I know some of my concerns are wider than just VT so specifically about the local RPM, could we at least keep the RPM more system-friendly by using the system-wide paths? You already mention it could customize paths on top so I see no reason why this shouldn't be possible.

Hi @beraldoleal, some feedback on your side is welcome too since you approved the changes after my comments.

**Comment by clebergnu** [comment]
> > At my first expectation when checking the modules under python's site-packages directory won't be about all possible files of the packages other than the actual importable modules. What is your opinion about this? I know some of my concerns are wider than just VT so specifically about the local RPM, could we at least keep the RPM more system-friendly by using the system-wide paths? You already mention it could customize paths on top so I see no reason why this shouldn't be possible.
> 
> Hi @beraldoleal, some feedback on your side is welcome too since you approved the changes after my comments.

Hi @pevogam, obviously I'm no @beraldoleal but I thought I'd let you know that I will reply to your comments, and won't merge this before we further discuss it.  Cheers!

**Comment by clebergnu** [comment]
> Hi @clebergnu, basic tests are ok and more are running now so I am ready to comment. All the changes here look completely reasonable but I am wondering something about the change in the data files path. In particular, the current standard python way seems to be:
> 
> 1. too monolithic placing various data files in the same location where the library code is placed
> 2. disregarding of system wide paths and common practices like using `/usr/share` for additional files and `/misc` for configuration for instance
> 3. created a clearer and clearer system-within-the-system feel of python packages being differently named and starting to use their own paths for nearly everything
> 

I agree with your perception here, and I don't blame Python itself for leaning towards a that approach.  It's really the distribution/packager to adapt it to a system's best practices.  In the past, in Avocado (non-VT) repo, we'd favor the upstream installation to be really a distribution specific layout, which caused problems when we tried to make "avocado-framework" easy to deploy by most users.  Then, we reversed (some) of it.  One example is the configuration files in the Avocado RPM, which use the `/etc/avocado` location. 

> At my first expectation when checking the modules under python's site-packages directory won't be about all possible files of the packages other than the actual importable modules. What is your opinion about this? 

I think this is what we came to expect if we look at it from the GNU/Linux distribution PoV.  But is not how modern Python attempts to solve the problem... If you do, say a `pip install django`, you'll see the number of data files that live under the `lib/site-packages/django/` tree.  For instance, on a venv `/tmp/v2`, the top-level files/dirs you'll find are:

```
    /tmp/v2/bin/django-admin
    /tmp/v2/bin/django-admin.py
    /tmp/v2/lib/python3.7/site-packages/Django-3.1.3.dist-info/*
    /tmp/v2/lib/python3.7/site-packages/django/*
```

> I know some of my concerns are wider than just VT so specifically about the local RPM, could we at least keep the RPM more system-friendly by using the system-wide paths? You already mention it could customize paths on top so I see no reason why this shouldn't be possible.

Sure, that's possible.  I was leaning towards following the Python structure in the RPM, and closing my eyes to this "regression" from the packaging perspective simply because it would mean less logic when trying to find the files (this is what we ended up doing in Avocado too).  Anyway, I'll give it a try, and if I don't find a can of worms being opened, I'll send a v3 with those improvements.

File: virttest/shared/scripts/virtio_console_guest.py
Comment (ldoktor) [inline virttest/shared/scripts/virtio_console_guest.py:116]: This change requires at least to change the `io-github-autotest-qemu/qemu/tests/virtio_console.py` as well, not sure if other tests as well...

File: virttest/data_dir.py
Comment (ldoktor) [inline virttest/data_dir.py:22]: I am really glad to see this change, Avocado-vt was waiting for it way too long. It will require a lot of testing, though...

File: virttest/data_dir.py
Comment (ldoktor) [inline virttest/data_dir.py:118]: The get_root_dir is fairly used in many locations, maybe we should keep the `ROOT_DIR` (or we could adjust the tests in most cases as they usually want shared/data dir instead)

File: virttest/data_dir.py
Comment (ldoktor) [inline virttest/data_dir.py:261]: given the root_dir lost most of it's significance we might just remove it from this listing, what do you think?

**Review by ldoktor — APPROVED** [review]
Thank you, @clebergnu, this was one big technical debt pending for so long. There is one little issue with regards to tp-qemu dependency that needs to be addressed. Given the helper is quite specific I think we don't need a backward compatible way, but we should merge the fixes together so master keeps working well.

As for @pevogam concerns, maintaining system-like layout is possible, but was always a source of a great pain. Python is established well enough and this is (almost) purely python package now so it makes sense to avoid trying to bend it and just use the standard locations and packaging. As Cleber mentioned there are still some places where we do want to extend this of our custom locations, but in general following the usual packaging should make avocado-vt less surprising to newcommers or users that are familiar with python.

Anyway let's wait for the merge with your feedback, in the mean-time, @clebergnu, please address the tp-qemu change and add a link here so they can be merged together.

File: virttest/data_dir.py
Comment (pevogam) [reply virttest/data_dir.py:22]: I am doing some large scale testing of our test suite and everything works well there so I assume follow-up errors will be restricted to tests and thus test suites mostly.

File: virttest/data_dir.py
Comment (pevogam) [reply virttest/data_dir.py:261]: It is still useful to print it though and this doesn't harm in any way or really increase complexity.

**Comment by pevogam** [comment]
Hi @clebergnu and @ldoktor, I used the last night to do some polling and collect more opinions. Here is what I could conclude:

> As for @pevogam concerns, maintaining system-like layout is possible, but was always a source of a great pain. Python is established well enough and this is (almost) purely python package now so it makes sense to avoid trying to bend it and just use the standard locations and packaging. As Cleber mentioned there are still some places where we do want to extend this of our custom locations, but in general following the usual packaging should make avocado-vt less surprising to newcommers or users that are familiar with python.

Right but in a way this contradicts the entire system-wide arrangement of files simply because the python project decided to put everything in one super-folder and call it "standard", essentially creating its own subsystem within the main system and thus taking a step away from manageability. I think you really understand the system administrator's pain here when one programming language decides that it is large enough to follow its own regulations and file structure. Does this mean one should do the same for Lua, Ruby, Rust, and so on then, creating a subsystem and standard locations for each? This is some monolithic behavior forming in a highly modular system such as linux and it is harming the ecosystem of multi-language packages in a way.

Ultimately it boils down to assigned responsibilities though. It is not the packaging developer's job to arrange these locations (which is the role we play here) and is a concern to address to the distribution packaging administrators. The developers should just focus on providing their package in the most platform-agnostic way available and the distro re-packager should focus on the system-specific ports. In this sense we actually make a step in the right direction and this PR is now all green for me. I am still happy that at least we raised some awareness about this here.

**Review by pevogam — APPROVED** [review]

**Comment by beraldoleal** [comment]
> > At my first expectation when checking the modules under python's site-packages directory won't be about all possible files of the packages other than the actual importable modules. What is your opinion about this? I know some of my concerns are wider than just VT so specifically about the local RPM, could we at least keep the RPM more system-friendly by using the system-wide paths? You already mention it could customize paths on top so I see no reason why this shouldn't be possible.
> 
> Hi @beraldoleal, some feedback on your side is welcome too since you approved the changes after my comments.

Hi @pevogam sorry, I missed your comment.

I understand your point. I'm used to the FHS too and I'm always trying to find files on the common standard UNIX directories. Since distutils, python packages are known to make Linux distributors unhappy. There is a lot of discussions about this in the Python community. But the truth is that each distro can choose different FHS layouts and still be following the FHS.

Because of that, it is not so easy to follow a standard system-wide layout and make your software available in many distros (just saying, not sure if we need this in this case).

Based on my experience, when distributing python packages we should try to follow the "python way" and separate what is a "python standard" from the FHS. Trying to distribute files outside the "python structure" will cause us some pain and extra work on the source code. Poluting the source code with some distro specific 'workarounds'.

Of course distros packages can handle this better and try to make the user's life easier. But this IMO should be a different step so each distro maintainer could handle this on the desired way. I'm not familiar with the Fedora rules when packaging some Python packages, but if we need this too, this could be an additional step and not the default one.

**Comment by pevogam** [comment]
Hi @beraldoleal,

I already concluded my part here but will leave some extra comments of your follow-up info:

> I understand your point. I'm used to the FHS too and I'm always trying to find files on the common standard UNIX directories. Since distutils, python packages are known to make Linux distributors unhappy. There is a lot of discussions about this in the Python community. But the truth is that each distro can choose different FHS layouts and still be following the FHS.

These are problems in the FHS that should eventually be addressed and fixed in the FHS and are not a sufficient reason for saying "we want to go the other way" as it doesn't fix the global issue but the local one.

> Based on my experience, when distributing python packages we should try to follow the "python way" and separate what is a "python standard" from the FHS. Trying to distribute files outside the "python structure" will cause us some pain and extra work on the source code. Polluting the source code with some distro-specific 'workarounds'.
> 
> Of course distros packages can handle this better and try to make the user's life easier. But this IMO should be a different step so each distro maintainer could handle this in the desired way. I'm not familiar with the Fedora rules when packaging some Python packages, but if we need this too, this could be an additional step and not the default one.

Yes, precisely what I said above about the separation of duties which often overlap when it comes to packaging and deployment.

> > "Ultimately it boils down to assigned responsibilities though. "
> 
> Yes, I agree with you here (speaking generally, in many packages out there), but regardless of who is going to do it, IMO should be an extra step, not the default.

The distros are doing good by trying to centralize packages from all sorts of languages of implementation and it is already a hard job because it really depends on the developers willingness to learn and follow particular standards. I guess some people just give up on this idea right away and start their own every-man-for-himself packagers as an alternative but this is going beyond the scope of the current discussion.

Local package, local problem, local solution is good enough here. Distro specific extensions like Debian or RPM package recipes could always be added as an extra step in a "packaging" folder or some other standard way. The same here applies for the existence of SPEC file in the Avocado VT project root which should then be moved to a distro-specific scope in order for this project's packages to be truly neutral. Also the changes I was referring to here were part of the RPM file which is distro specific but let's simply move to distro neutrality and consider limiting the important of the SPEC file as well.

File: virttest/shared/scripts/virtio_console_guest.py
Comment (clebergnu) [reply virttest/shared/scripts/virtio_console_guest.py:116]: Good point! Here it is: https://github.com/autotest/tp-qemu/pull/2485

**Comment by clebergnu** [comment]
Pushing a rebased (unchanged) version to solve conflicts and trigger CI again.

**Comment by yanan-fu** [comment]
@clebergnu  The commit ba632749d62355ffda3b48f64c70f2f8135433f6 cause the file `virttest/passfd.c` can't be installed, and nic hotplug cases will failed. Could you help check it ?

Do you plan to add it back ?  or replace it with `socket.sendmsg` ?  For repIacement,  there is one existing patch: https://github.com/avocado-framework/avocado-vt/pull/2470.
Thanks!

**Comment by clebergnu** [comment]
> @clebergnu The commit [ba63274](https://github.com/avocado-framework/avocado-vt/commit/ba632749d62355ffda3b48f64c70f2f8135433f6) cause the file `virttest/passfd.c` can't be installed, and nic hotplug cases will failed. Could you help check it ?
> 

Yes, you're right that `passfd.c` ends up not being packaged after this change.  Sorry about that.

PR https://github.com/avocado-framework/avocado-vt/pull/2829 addresses that.

> Do you plan to add it back ? or replace it with `socket.sendmsg` ? For repIacement, there is one existing patch: #2470.
> Thanks!

We can continue the discussion on #2470 indeed.  In the mean time, #2829 is a hotfix to restore the passfd functionality.

**Pull request merged**
